File: demos/php/dsa/sorting/quick_sort.php
<?php
function quickSort(array $arr): array {
    if (count($arr) <= 1) return $arr;
    $pivot = $arr[intdiv(count($arr), 2)];
    $left = [];
    $equal = [];
    $right = [];
    foreach ($arr as $v) {
        if ($v < $pivot) $left[] = $v;
        elseif ($v > $pivot) $right[] = $v;
        else $equal[] = $v;
    }
    return array_merge(quickSort($left), $equal, quickSort($right));
}

echo "=== PHP Functional QuickSort ===\n";
$data = [64, 25, 12, 22, 11];
$sorted = quickSort($data);
assert($sorted === [11, 12, 22, 25, 64]);
echo "Sorted: " . implode(", ", $sorted) . "\n";
echo "PHP QuickSort tests passed successfully.\n";

File: demos/ruby/dsa/sorting/quick_sort.rb
def quick_sort(arr)
  return arr if arr.length <= 1
  pivot = arr[arr.length / 2]
  left = arr.select { |x| x < pivot }
  equal = arr.select { |x| x == pivot }
  right = arr.select { |x| x > pivot }
  quick_sort(left) + equal + quick_sort(right)
end

puts "=== Ruby Functional QuickSort ==="
data = [64, 25, 12, 22, 11]
sorted = quick_sort(data)
raise "Sort error" unless sorted == [11, 12, 22, 25, 64]

puts "Sorted: #{sorted}"
puts "Ruby QuickSort tests passed successfully."
